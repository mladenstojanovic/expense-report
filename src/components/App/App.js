import React from 'react';
import { useSelector } from 'react-redux';
import CreateUser from '../CreateUser/CreateUser';
import NetworkStatus from '../NetworkStatus/NetworkStatus';
import { AppHeaderStyle, AppLogoStyle } from './App.style';
import {
  IDLE,
  IN_PROGRESS
} from '../../store/actions/network/network.constants';
import TransactionData from '../TransactionData/TransactionData';

const App = () => {
  const networkStatus = useSelector(state => state.network.networkStatus);
  return (
    <AppHeaderStyle>
      <AppLogoStyle
        src={'https://basiq.io/a0a4ce023263c9b295846d60914ed319.svg'}
        className="App-logo"
        alt="logo"
      />
      {networkStatus === IDLE ? (
        <CreateUser />
      ) : networkStatus === IN_PROGRESS ? (
        <NetworkStatus />
      ) : (
        <TransactionData />
      )}
    </AppHeaderStyle>
  );
};

export default App;
