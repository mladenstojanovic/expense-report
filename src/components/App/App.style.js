import styled from 'styled-components/macro';

export const AppHeaderStyle = styled.div`
  background-color: ${props => props.theme.colors.darkBlue};
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
`;

export const AppLogoStyle = styled.img`
  height: 40vmin;
`;
