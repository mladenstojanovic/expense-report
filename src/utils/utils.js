import React from 'react';
import { render } from '@testing-library/react';
import { createStore, combineReducers } from 'redux';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import networkReducer from '../store/reducers/network.reducer';
import theme from '../styles/theme';

export const sleep = milliseconds => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const sortTransactionData = data => {
  const sortedData = {};
  for (let index = 0; index < data.length; index++) {
    if (!data[index].subClass) {
      continue;
    }
    if (!sortedData[data[index].subClass.code]) {
      sortedData[data[index].subClass.code] = {
        total: 0,
        numOfTransactions: 0,
        title: data[index].subClass.title
      };
    }
    sortedData[data[index].subClass.code].total += Math.abs(
      parseInt(data[index].amount)
    );
    sortedData[data[index].subClass.code].numOfTransactions++;
  }
  return sortedData;
};

export const isEmailValid = email => /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);

export const isPhoneNumberValid = phone =>
  /^[+]?[(]?[0-9]{3}[)]?[-\s.]?[0-9]{3}[-\s.]?[0-9]{4,6}$/.test(phone);

export const renderWithReduxAndStyles = (
  ui,
  {
    initialState,
    store = createStore(
      combineReducers({ network: networkReducer }),
      initialState
    )
  } = {}
) => ({
  ...render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Provider>
  ),
  store
});
