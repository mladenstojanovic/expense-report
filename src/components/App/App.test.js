import React from 'react';
import App from './App';
import {
  IN_PROGRESS,
  SUCCESS
} from '../../store/actions/network/network.constants';
import { renderWithReduxAndStyles } from '../../utils/utils';

jest.mock('../CreateUser/CreateUser', () => ({
  __esModule: true,
  default: () => <p>Mock Create User</p>
}));

jest.mock('../NetworkStatus/NetworkStatus', () => ({
  __esModule: true,
  default: () => <p>Mock Network Status</p>
}));

jest.mock('../TransactionData/TransactionData', () => ({
  __esModule: true,
  default: () => <p>Mock Transaction Data</p>
}));

describe('Test App component', () => {
  it('should render correctly', () => {
    const { container } = renderWithReduxAndStyles(<App />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render create user component with default network status', () => {
    const { getByText } = renderWithReduxAndStyles(<App />);

    const mockChild = getByText('Mock Create User');

    expect(mockChild).toBeTruthy();
  });

  it('should render network status component with network status in progress', () => {
    const { getByText } = renderWithReduxAndStyles(<App />, {
      initialState: {
        network: {
          networkStatus: IN_PROGRESS
        }
      }
    });

    const mockChild = getByText('Mock Network Status');

    expect(mockChild).toBeTruthy();
  });

  it('should render transaction data component with network status success', () => {
    const { getByText } = renderWithReduxAndStyles(<App />, {
      initialState: {
        network: {
          networkStatus: SUCCESS
        }
      }
    });

    const mockChild = getByText('Mock Transaction Data');

    expect(mockChild).toBeTruthy();
  });
});
