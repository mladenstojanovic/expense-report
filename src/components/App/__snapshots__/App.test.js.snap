// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App component should render correctly 1`] = `
.c0 {
  background-color: #282c34;
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.c1 {
  height: 15vmin;
}

<div
  class="c0"
>
  <img
    alt="logo"
    class="c1 App-logo"
    src="https://basiq.io/a0a4ce023263c9b295846d60914ed319.svg"
  />
  <p>
    Mock Create User
  </p>
</div>
`;
